### DISCUSSION TITLE
Atwinc Driver Discussion

### DISCUSSION THREAD

**Discussion by DrewTChrist** [DrewTChrist/atwinc1500-rs#15] [General]
Let's talk about the driver!

**Comment by owlshrimp** [comment]
I gather that at this stage you've pretty much got communication over SPI figured out, and you're working on telling the atwinc to connect to a station?

**Reply by DrewTChrist** [reply]
For the most part yes. The data sheet defines some error recovery mechanisms which I'd like to implement for SPI communication. Without them in place I'm sure the driver won't work as expected for more complex tasks.

Connecting to a station is possible right now for wpa2 for wincs having firmware lower than 19.7.6. They changed the connection format in 19.7.6 and deprecated WEP so I plan on handling those cases accordingly.

Once those things are finished I'd say the next thing that needs to be done is implementing a callback system similar to the one in the atmel driver. That portion has to be in place before any of the socket stuff, and much of the other Atwinc functionality (grabbing system time or RSSI).

**Reply by owlshrimp** [reply]
Right now you're uploading different versions of the firmware using a regular arduino program (not in rust) I guess?
What dev board/MCU are you currently using? An AVR?

**Reply by DrewTChrist** [reply]
My only Atwinc has firmware version 19.5.2 I believe and I have not tried to update it as I don't think there is any way to downgrade the firmware after that. I could be wrong on that.

I've been using a Raspberry Pi Pico for development, but I've also been testing the driver on different targets in this repo [here](https://github.com/drewtchrist/atwinc1500-rs-examples/tree/develop).

Edit: It would be helpful for me to know what kind of Atwinc1500 board you are using. There are probably many designs out there, I'm using the Adafruit breakout board.

**Reply by owlshrimp** [reply]
I have several modules, but they're the bare module and I have yet to design my own breakout. I have studied the adafruit design though. While they will probably feature an atmega168pv (1KB SRAM) for testing, I'll include testpoints so that I can connect a different MCU later (eg. a atmega1280pv with 8KB or a Pico).

**Reply by DrewTChrist** [reply]
I see. Well that's good to know. I've been wondering lately if most people are using Adafruit boards or custom designs.

I'm currently focusing on figuring out the licensing situation for the driver. Because the data sheet does not provide enough information to write a driver, I've been working with the old Atmel C drivers. Well since Microchip bought them they've relicensed everything under their own restrictive license that doesn't play well with open source licenses or "derivatives". I'm not sure if that same license then legally applies to the older versions as well. I don't want to get into any legal issues, but I'd like this project to be as open and usable as possible for anyone interested. Here is the [current license](https://github.com/Microchip-MPLAB-Harmony/wireless_wifi/blob/master/mplab_harmony_license.md) and here is the [older license header](https://github.com/arduino-libraries/WiFi101/blob/master/src/driver/source/m2m_hif.c) which I think is Apache 1.1.

**Reply by owlshrimp** [reply]
IANAL, etc, etc, but:

1) It seems that arduino are continuing to distribute the old code under the old licence (at the link you provide), so I wonder how possible it actually is for atmel/microchip to change things this late after things have been distributed around? eg. if Arduino have modified the code and distributed it under the same relatively permissive licence.

2) There seems to be some agreeing opinions online that the MIT licence (which seems to be the most similar to the old terms) can't be revoked, or at least can only be changed for new versions of the software. The new licence seems to be applied to code as bundled in Microchip's MPLab Harmony.

It would be interesting to see how many organizations are still distributing the old code under the old terms, or even if there's a place microchip still is.

3) While you're certainly learning from their code, so it's not a total cleanroom, there is the question in my mind of how truly similar your code is or if it's merely "inspired."

**Reply by DrewTChrist** [reply]
I read and asked around. It seems the consensus is once that version of the code has been released under a certain license it will always be. They can change the license for future versions as you said.

Also I did take the crc syndrome table in crc.rs directly from the Atmel driver. I'm not sure how those are made, but I don't believe that poses an issue in this case either.

I pushed a much needed release so the crate is up with the docs. It's still not very functional though.

**Reply by owlshrimp** [reply]
What is the CRC table used for? Is it just a list of constants used for computing the CRC?

**Reply by DrewTChrist** [reply]
Yeah just a huge array of values that somehow help calculate the CRC7 for sending SPI commands.

**Reply by owlshrimp** [reply]
I wouldn't be hugely worried about anything like that which has to be a particular way (ie. have to have particular values) in order for it to work.

**Comment by owlshrimp** [comment]
How much is handled on the atwinc, and how much on the MCU? Like do you tell it to make a connection and exchange ethernet frames with it, then implement TCP/IP (and TCP/UDP) yourself? Or does it give you TCP streams and UDP datagrams?

**Reply by DrewTChrist** [reply]
The atwinc handles everything essentially. The host driver really only needs to maintain the state between them such as what connections have been made or how many sockets are in use. I believe it gives the TCP streams and UDP datagrams and I don't think there is any exchange of ethernet frames between the atwinc and the host mcu. I plan to implement the traits in [embedded-nal](https://docs.rs/embedded-nal/latest/embedded_nal/) eventually which pretty much define the API.

**Reply by owlshrimp** [reply]
Does the atwinc handle TLS, eg. for HTTPS connections, or is that left to the MCU firmware on top of the TCP connection? I could imagine the former being less-than-optimal, being as we'd need to wait for atmel to ship a patch if there's a vulnerability.

**Reply by DrewTChrist** [reply]
Yes the atwinc does handle TLS so firmware updates would be needed to get any fixes like you said. According to [this](https://learn.adafruit.com/adafruit-atwinc1500-wifi-module-breakout/updating-ssl-certificates) some modules will ship with support for 1.2. The firmware change logs probably specify which firmware supports TLS 1.2 and it's probably one of the relatively newer versions. So it's already behind.
